### PR TITLE
Added SVG Map as-is from mapsvg.com

### PR DESCRIPTION
**Pull request by BillWu-345**
Added SVG Map

### CODE REVIEW

**Comment by KeweiCodes** [comment]
So I think we should store all SVG path info in a TS file instead, and render the SVG through code. 
See: https://github.com/StephanWagner/svgMap/blob/master/src/js/svgMap.js#L1342

This would keep our JS module purely JS, which makes it easier to import for other users. 
It also gives us flexibility in the future to do things like render only specified regions.

**Comment by KeweiCodes** [comment]
Btw, the index.html file is for our development purpose only, in the future when we publish this module for others to use, the html file won't be included.

**Comment by BillWu-345** [comment]
Done! I'm just a bit sloppy on the type declarations.

File: src/index.ts [hunk 1]
```diff
@@ -1,31 +1,144 @@
...
+    }
+  };
 }
 
 
 // Example usage
 const options = {
+  targetElement: "map",
```
Comment (KeweiCodes) [inline src/index.ts]: Can we rename this option to something like "targetSelector", and then we use it as the input for `document.querySelector` instead? 

https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector

File: src/index.ts [hunk 2]
```diff
@@ -1,31 +1,144 @@
...
+        countryElement.setAttribute('data-id', countryID);
+        countryElement.classList.add('svgMap-country');
+
+        this.mapImage.appendChild(countryElement);
+      }.bind(this)
+    );
+  }
+
+  error(error) {
+    (console.error || console.log)(
```
Comment (KeweiCodes) [inline src/index.ts]: I think it's better to throw an error here.

```ts
throw new Error(`svgMap error: ${error || 'Unknown error'}`)
```

In this case if we can't find the target element, we definitely don't care about any further execution in our code so we might as well throw it so the user can at least notice and handle it.
https://stackoverflow.com/a/60383917/15828150

File: src/index.html [hunk 3]
```diff
@@ -6,6 +6,7 @@
     <meta name="viewport" content="width=device-width, initial-scale=1" />
   </head>
   <body>
-    <div id="map">Map goes here?</div>
+    <div id="map"></div>
+    <?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (KeweiCodes) [inline src/index.html]: What's this for?

File: src/index.ts [hunk 4]
```diff
@@ -1,31 +1,144 @@
...
+
+        var countryElement = document.createElementNS(
+          'http://www.w3.org/2000/svg',
+          'path'
+        );
+
+        countryElement.setAttribute('d', countryData.d);
+        countryElement.setAttribute(
+          'id',
+          this.id + '-map-country-' + countryID
```
Comment (KeweiCodes) [inline src/index.ts]: What's this.id? Doesn't look it exists?

Why can't we just use `countryID` here as the id value?

File: src/index.ts [hunk 5]
```diff
@@ -1,31 +1,144 @@
...
+    this.mapImage.setAttribute('viewBox', '0 0 701.42358 988.42462');
+    this.mapImage.setAttribute('fill', 'white');
+    this.mapImage.setAttribute('stroke', 'black');
+    this.mapImage.classList.add('svgMap-map-image');
+
+
+    // <path> inside <svg>
+    Object.keys(this.mapPaths).forEach(
+      function (countryID) {
+        var countryData = this.mapPaths[countryID];
```
Comment (KeweiCodes) [inline src/index.ts]: regionData

File: src/index.ts [hunk 6]
```diff
@@ -1,31 +1,144 @@
...
+        if (!countryData.d) {
+          return;
+        }
+
+        var countryElement = document.createElementNS(
```
Comment (KeweiCodes) [inline src/index.ts]: regionElement

File: src/index.ts (hunk 1 above)
Comment (BillWu-345) [reply src/index.ts]: done

File: src/index.html (hunk 3 above)
Comment (BillWu-345) [reply src/index.html]: yup line deleted, was obsolete code from the nzsvg website

File: src/index.ts (hunk 4 above)
Comment (BillWu-345) [reply src/index.ts]: yup this was already changed in my latest version that I haven't pushed up yet

File: src/index.ts (hunk 5 above)
Comment (BillWu-345) [reply src/index.ts]: yup this was already changed in my latest version that I haven't pushed up yet

File: src/index.ts (hunk 6 above)
Comment (BillWu-345) [reply src/index.ts]: yup this was already changed in my latest version that I haven't pushed up yet

File: src/index.ts [hunk 7]
```diff
@@ -1,31 +1,181 @@
...
+    'AUK': "Thinks rest of NZ doesn't exist",
+    'STL': "100km/hr, is that the minimum speed?",
+    'OTA': "There are summer clothes, and.. summer clothes",
+    'WGN': "Everyone I know lives in a shed",
+    'WKO': "The Secret Cow Level in Diablo was inspired by us",
+    'BOP': "Too many retired people from Auckland",
+    'GIS': "How do I get a job? Lets go to Auckland",
+    'HKB': "Money don't grow on trees. Nope but apples do, and apples is $$",
+    'NTL': "The rest of NZ is too cold",
+    'CAN': "We're NZ's biggest city, Auckland doesn't count"
```
Comment (KeweiCodes) [inline src/index.ts:171]: 😂

File: src/index.ts [hunk 8]
```diff
@@ -1,31 +1,181 @@
...
+      regionElement.setAttribute('style', {'color': this.options.colors[regionID]});
+      this.mapImage.appendChild(regionElement);
+
+      // Test for hover code:
+      regionElement.addEventListener(
+        'mouseover',
+        (e) => {
+          let regionID = e.target.id;
+          let regionDescription = this.options.descriptions[regionID];
+          console.log(regionID +': '+ regionDescription);
```
Comment (KeweiCodes) [inline src/index.ts]: Remove this.

File: src/index.ts [hunk 9]
```diff
@@ -1,31 +1,181 @@
...
+          this.descriptionDisplay.innerHTML = regionID +': '+ regionDescription
```
Comment (KeweiCodes) [inline src/index.ts]: We should be weary whenever we try set a string to HTML through innerHTML, we can use textContent instead, see: https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent

The main difference is textContent just renders the string as text which prevents XSS attacks.
E.g. a malicious chrome extension might set the region id tag to `<iframe src="mybank.com/transfer?to=kewei&amount=10000"/>`, which could have bad consequences if the browser rendered that as HTML.

File: src/index.ts [hunk 10]
```diff
@@ -1,31 +1,181 @@
...
+        }
+      );
+    }
+  }
+
+  error(error) {
+    throw new Error(`svgMap error: ${error || 'Unknown error'}`);
```
Comment (KeweiCodes) [inline src/index.ts]: Is this missing a `}`?

File: src/index.ts [hunk 11]
```diff
@@ -1,31 +1,181 @@
...
   constructor(options: MapOptions) {
-    this.options = options
+    this.options = options;
   }
 
   // Add other methods here
+  render() {
+    if (
+      !this.options.targetSelector ||
+      !document.querySelector(`#${this.options.targetSelector}`)
```
Comment (KeweiCodes) [inline src/index.ts:38]: Just do `document.querySelector(this.options.targetSelector})`

You can pass in any css selector to `document.querySelector `

Comment (KeweiCodes) [reply src/index.ts:38]: E.g. you can pass in 
{ ...targetSelector: "#map" }
or 
{ ...targetSelector: ".container > .map" }
or 
{ ...targetSelector: "div.map" } etc.
https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector

Comment (BillWu-345) [reply src/index.ts:38]: so this.options.targetSelector == map
`#${this.options.targetSelector}` == #map

the first one doesn't work

File: src/index.ts (hunk 10 above)
Comment (BillWu-345) [reply src/index.ts]: Yep it was, now fixed

File: src/index.ts (hunk 9 above)
Comment (BillWu-345) [reply src/index.ts]: ok changed

File: src/index.ts (hunk 8 above)
Comment (BillWu-345) [reply src/index.ts]: done

File: src/index.ts (hunk 7 above)
Comment (BillWu-345) [reply src/index.ts:171]: 😂

**Review by KeweiCodes — APPROVED** [review]

**Pull request merged**